# single-cell RNA-seq and spatial transcriptomics analysis pipeline

The primary objective of this study is to gain hands-on experience with the scRNA-seq and spatial transcriptomics workflow, including data preprocessing, normalization, clustering, and downstream analysis. All analytical processes in this study are implemented using R/Python based on the updated related packages.

The dataset were downloaded from a research study by Zhu et al. (2022), titled "Delineating the dynamic evolution from preneoplasia to invasive lung adenocarcinoma by integrating single-cell RNA sequencing and spatial transcriptomics"(https://doi.org/10.1038/s12276-022-00896-9). Nine primary LUAD samples were collected from nine patients at three stages, including 3 adenocarcinoma in situ (AIS), 3 minimally invasive adenocarcinoma (MIA), and 3 invasive adenocarcinoma (IAC). 

The datasets includes:

1),ScRNA-seq using NovaSeq 6000 (Illumina) platforms: A total of 115,246 cells consists of 37,143 cells (32.2%) from AIS, 30,909 cells (26.8%) from MIA, and 47,194 cells (41.0%) from IAC through scRNA-seq. 

2),Spatial RNA-seq using 10Ã— Genomics Visium platforms: six specimens (AIS: TD5, TD8; MIA: TD3, TD6;and IAC: TD1, TD2) analyzed by scRNA-seq were further subjected to ST.

